src/Markup/Avalonia.Markup/Data/Binding.cs

```diff
@@ -1,208 +1,18 @@
-                nodes = ExpressionNodeFactory.CreateFromAst(
-                    astPool,
-                    TypeResolver,
-                    GetNameScope(),
-                    out isRooted);
-            }
-
-            // If the binding isn't rooted (i.e. doesn't have a Source or start with $parent, $self,
-            // #elementName etc.) then we need to add a source node. The type of source node will
-            // depend on the ElementName and RelativeSource properties of the binding and if
-            // neither of those are set will default to a data context node.
-            if (Source == AvaloniaProperty.UnsetValue && !isRooted && CreateSourceNode(targetProperty) is { } sourceNode)
-            {
-                nodes ??= new();
-                nodes.Insert(0, sourceNode);
-            }
-
-            // If the first node is an ISourceNode then allow it to select the source; otherwise
-            // use the binding source if specified, falling back to the target.
-            var source = nodes?.Count > 0 && nodes[0] is SourceNode sn ?
-                sn.SelectSource(Source, target, anchor ?? DefaultAnchor?.Target) :
-                Source != AvaloniaProperty.UnsetValue ? Source : target;
-
-            var (mode, trigger) = ResolveDefaultsFromMetadata(target, targetProperty);
-
-            return new BindingExpression(
-                source,
-                nodes,
-                FallbackValue,
-                delay: TimeSpan.FromMilliseconds(Delay),
-                converter: Converter,
-                converterCulture: ConverterCulture,
-                converterParameter: ConverterParameter,
-                enableDataValidation: enableDataValidation,
-                mode: mode,
-                priority: Priority,
-                stringFormat: StringFormat,
-                targetProperty: targetProperty,
-                targetNullValue: TargetNullValue,
-                targetTypeConverter: TargetTypeConverter.GetReflectionConverter(),
-                updateSourceTrigger: trigger);
-        }
-
-        private INameScope? GetNameScope()
-        {
-            INameScope? result = null;
-            NameScope?.TryGetTarget(out result);
-            return result;
-        }
-
-        private ExpressionNode? CreateSourceNode(AvaloniaProperty? targetProperty)
-        {
-            if (!string.IsNullOrEmpty(ElementName))
-            {
-                var nameScope = GetNameScope() ?? throw new InvalidOperationException(
-                    "Cannot create ElementName binding when NameScope is null");
-                return new NamedElementNode(nameScope, ElementName);
-            }
+/// <summary>
+/// Provides limited compatibility with the 11.x Binding class. Use <see cref="ReflectionBinding"/>
+/// for new code.
+/// </summary>
+[RequiresUnreferencedCode(TrimmingMessages.ReflectionBindingRequiresUnreferencedCodeMessage)]
+public class Binding : ReflectionBinding
```

Comment:
Side note after having seen everything else: it might also be worth considering simply renaming `BindingBase` to `Binding`.

Reply:
Not sure: `BindingBase` makes it more apparent that you must use a derived class.